### PR TITLE
*: support cgroup with systemd

### LINKED ISSUE #47442: tidb cannot read resource limitation from cgroup with the systemd
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 2. What did you expect to see? (Required)

can read the resource limit

### 3. What did you see instead (Required)

cannot read the resource limit

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #47442

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

1、with systemd 

run tidb-server 

```
[Unit]
Description=tidb

[Service]
Type=simple
Restart=always
ExecStart=/usr/bin/tidb
StandardOutput=journal
MemoryMax=500M

[Install]
WantedBy=multi-user.target
```

you will see the log like this.

```
 [2023/10/30 11:00:49.325 +00:00] [INFO] [meminfo.go:188] ["use cgroup memory hook"] [cgroupMemorySize=524288000] [physicalMemorySize=32900468736]
```

2、without systemd

you will see.

```
[2023/10/30 10:47:16.154 +00:00] [INFO] [meminfo.go:190] ["use physical memory hook"] [cgroupMemorySize=9223372036854775807] [physicalMemorySize=32900468736]
```

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
*: support cgroup with systemd
```

### CODE REVIEW

**Review by tiancaiamao — APPROVED** [review]

File: pkg/util/memory/meminfo.go [hunk 1]
```diff
@@ -163,6 +170,37 @@ func init() {
 	terror.MustNil(err)
 }
 
+// InitMemoryHook initializes the memory hook.
+// It is to solve the problem that tidb cannot read cgroup in the systemd.
+// so if we are not in the container, we compare the cgroup memory limit and the physical memory,
+// the cgroup memory limit is smaller, we use the cgroup memory hook.
+func InitMemoryHook() {
+	if !cgroup.InContainer() {
```
Comment (YangKeao) [inline pkg/util/memory/meminfo.go:178]: How about `if cgroup.InContainer() {logxxxx; return}`?

**Review by YangKeao — COMMENTED** [review]
Rest LGTM

File: pkg/util/memory/meminfo.go (hunk 1 above)
Comment (hawkingrei) [reply pkg/util/memory/meminfo.go]: Done

File: pkg/util/memory/meminfo.go [hunk 2]
```diff
@@ -163,6 +170,37 @@ func init() {
 	terror.MustNil(err)
 }
 
+// InitMemoryHook initializes the memory hook.
+// It is to solve the problem that tidb cannot read cgroup in the systemd.
+// so if we are not in the container, we compare the cgroup memory limit and the physical memory,
+// the cgroup memory limit is smaller, we use the cgroup memory hook.
+func InitMemoryHook() {
+	if cgroup.InContainer() {
+		logutil.BgLogger().Info("use cgroup memory hook because TiDB is in the docker.")
```
Comment (hi-rustin) [inline pkg/util/memory/meminfo.go:179]:
```suggestion
		logutil.BgLogger().Info("Use cgroup memory hook because TiDB is in the docker.")
```

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/util/memory/meminfo.go [hunk 3]
```diff
@@ -163,6 +170,37 @@ func init() {
 	terror.MustNil(err)
 }
 
+// InitMemoryHook initializes the memory hook.
+// It is to solve the problem that tidb cannot read cgroup in the systemd.
+// so if we are not in the container, we compare the cgroup memory limit and the physical memory,
+// the cgroup memory limit is smaller, we use the cgroup memory hook.
+func InitMemoryHook() {
```
Comment (hi-rustin) [inline pkg/util/memory/meminfo.go:177]: Can I ask why we need to have an init hook and an InitMemoryHook at the same time? Can we only put it into the explicit API or in the init function? Then we don't need to update the value in the different stages and places.

**Review by hi-rustin — COMMENTED** [review]
Thanks!

File: pkg/util/memory/meminfo.go (hunk 3 above)
Comment (hawkingrei) [reply pkg/util/memory/meminfo.go:177]: ```InitMemoryHook``` need to print log. but we cannot write log in the ```init```. logging have been inited. so we must run ```initMemoryHook``` after the logging's init.

Comment (hi-rustin) [reply pkg/util/memory/meminfo.go:177]: So maybe we can move all of them into InitMemoryHook.

**Pull request merged**